File: aws/cloudformation/cfn_response.rb
require 'net/http'
require 'json'
require 'uri'

module CfnResponse
  def self.send(
    event,
    context,
    response_status = 'SUCCESS',
    response_data: {},
    physical_resource_id: nil,
    message: nil
  )
    response_body = {
      Status: response_status,
      Reason: message || "See the details in CloudWatch Log Stream: " + context.log_stream_name,
      PhysicalResourceId: physical_resource_id,
      StackId: event['StackId'],
      RequestId: event['RequestId'],
      LogicalResourceId: event['LogicalResourceId'],
      Data: response_data
    }.to_json

    puts "Response body:\n#{response_body}"

    url = event['ResponseURL']
    uri = URI(url)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true if uri.scheme == 'https'
    http.open_timeout = http.read_timeout = 30

    req = Net::HTTP::Put.new(url)
    req.body = response_body
    req.content_length = response_body.bytesize

    res = http.request(req)
    [res.code, res.each_header.to_h, res.body]
  end
end

File: aws/cloudformation/lambdas/inline/cfn-response.js
// This is a local copy of the cfn-response module.
//
// When you use the ZipFile property to specify your function's source code and
// that function interacts with an AWS CloudFormation custom resource, you can
// load the cfn-response module to send responses to those resources. You must
// include the module via `var response = require('cfn-response');` in order for
// CloudFormation to correctly include the module.
//
// See: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-lambda-function-code-cfnresponsemodule.html

exports.SUCCESS = "SUCCESS";
exports.FAILED = "FAILED";

exports.send = function(event, context, responseStatus, responseData, physicalResourceId, message) {
  physicalResourceId = physicalResourceId || context.logStreamName;

  // If this lambda function does not have permissions to CloudWatch Logs
  // a PhysicalResourceId is still needed to send to CloudFormation.
  if (event.RequestType === 'Create' && responseStatus === exports.FAILED) {
    physicalResourceId = physicalResourceId || exports.FAILED;
  }

  var responseBody = JSON.stringify({
    Status: responseStatus,
    Reason: message || "See the details in CloudWatch Log Stream: " + context.logStreamName,
    PhysicalResourceId: physicalResourceId,
    StackId: event.StackId,
    RequestId: event.RequestId,
    LogicalResourceId: event.LogicalResourceId,
    Data: responseData
  });

  console.log("Response body:\n", responseBody);

  var https = require("https");
  var url = require("url");

  var parsedUrl = url.parse(event.ResponseURL);
  var options = {
    hostname: parsedUrl.hostname,
    port: 443,
    path: parsedUrl.path,
    method: "PUT",
    headers: {
      "content-type": "",
      "content-length": responseBody.length
    }
  };

  var request = https.request(options, function(response) {
    console.log("Status code: " + response.statusCode);
    console.log("Status message: " + response.statusMessage);
    context.done();
  });

  request.on("error", function(error) {
    console.log("send(..) failed executing https.request(..): " + error);
    context.done();
  });

  request.write(responseBody);
  request.end();
};
